tools/c7n_kube/tests/test_k8s_server.py

```diff
@@ -6,87 +6,82 @@
 import threading
 import time
 
+from contextlib import contextmanager
+
 import requests
 
 from unittest.mock import patch, MagicMock
 
-from c7n_kube.server import \
-    AdmissionControllerServer, AdmissionControllerHandler, init
+from c7n_kube.server import AdmissionControllerServer, AdmissionControllerHandler, init
 
 from common_kube import KubeTest
 
 
 class TestAdmissionControllerServer(AdmissionControllerServer):
     def __init__(self, bind_and_activate=False, *args, **kwargs):
-        super().__init__(
-            bind_and_activate=bind_and_activate, *args, **kwargs)
+        super().__init__(bind_and_activate=bind_and_activate, *args, **kwargs)
 
 
 class TestServer(KubeTest):
-
     def find_port(self):
         sock = socket.socket()
-        sock.bind(('', 0))
-        _, port = sock.getsockname()
+        sock.bind(("", 0))
+        ((server, port)) = sock.getsockname()
         return port
 
-    def _server(self, policies, on_exception='warn'):
+    @contextmanager
+    def _server(self, policies, on_exception="warn", lifetime=1):
         port = self.find_port()
         with tempfile.TemporaryDirectory() as temp_dir:
             with open(f"{temp_dir}/policy.yaml", "w+") as f:
                 json.dump(policies, f)
             server = TestAdmissionControllerServer(
-                server_address=('localhost', port),
+                server_address=("localhost", port),
                 RequestHandlerClass=AdmissionControllerHandler,
                 policy_dir=temp_dir,
                 bind_and_activate=True,
                 on_exception=on_exception,
             )
-            server_thread = threading.Thread(target=server.handle_request)
+            server_thread = threading.Thread(target=server.serve_forever)
             server_thread.start()
-            time.sleep(1)
-        return server, port
+            time.sleep(lifetime)
+        try:
+            yield server, port
+        finally:
+            server.shutdown()
```

Comment:
add server_thread.join() else the actual thread shutdown happens asynchronously outside of the test. server_forever is polling on a socket with timeout, till the timeout is reached the shutdown isn't processed.